src/Umbraco.Core/Services/LocalizationService.cs

```diff
@@ -453,26 +454,103 @@ public IEnumerable<ILanguage> GetAllLanguages()
     /// <param name="language"><see cref="ILanguage" /> to save</param>
     /// <param name="userId">Optional id of the user saving the language</param>
     public void Save(ILanguage language, int userId = Constants.Security.SuperUserId)
+    {
+        Attempt<ILanguage, LanguageOperationStatus> result = language.Id > 0
+            ? Update(language, userId)
+            : Create(language, userId);
+
+        // mimic old Save behavior
+        if (result.Status == LanguageOperationStatus.InvalidFallback)
+        {
+            throw new InvalidOperationException($"Cannot save language {language.IsoCode} with fallback id={language.FallbackLanguageId}.");
+        }
+    }
+
+    /// <inheritdoc />
+    public Attempt<ILanguage, LanguageOperationStatus> Create(ILanguage language, int userId = Constants.Security.SuperUserId)
+    {
+        using (ICoreScope scope = ScopeProvider.CreateCoreScope())
+        {
+            // write-lock languages to guard against race conds when dealing with default language
+            scope.WriteLock(Constants.Locks.Languages);
+
+            if (HasValidIsoCode(language) == false)
+            {
+                return Attempt.FailWithStatus(LanguageOperationStatus.InvalidIsoCode, language);
+            }
+
+            // ensure no duplicates by ISO code
+            if (_languageRepository.GetByIsoCode(language.IsoCode) != null)
+            {
+                return Attempt.FailWithStatus(LanguageOperationStatus.DuplicateIsoCode, language);
+            }
+
+            // ensure valid fallback language (note: new languages cannot create cycles, no need to check for that)
+            if (language.FallbackLanguageId.HasValue && _languageRepository.Exists(language.FallbackLanguageId.Value) == false)
+            {
+                return Attempt.FailWithStatus(LanguageOperationStatus.InvalidFallback, language);
+            }
+
+            EventMessages eventMessages = EventMessagesFactory.Get();
+            var savingNotification = new LanguageSavingNotification(language, eventMessages);
+            if (scope.Notifications.PublishCancelable(savingNotification))
+            {
+                scope.Complete();
+                return Attempt.FailWithStatus(LanguageOperationStatus.CancelledByNotification, language);
+            }
+
+            // explicitly ensure a new language is created in case of model reuse
+            language.Id = 0;
```

Comment:
I wonder if it wouldn't be better to fail the operation here? Instead of trying to assume the intention behind passing in a language with an ID?

Reply:
True that. In this particular case it's not super important, but as a general rule it is... I'll amend the PR to match 😄